### PR TITLE
Implement WireframeError error integration (#513)

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch implements standard error integration for `WireframeError` in support of issue #513. It makes the default public error type usable through `std::error::Error`, preserves source chaining for transport and codec failures, and lets `TestError` use thiserror's transparent `#[from]` conversion.

Closes #513.

## Review walkthrough

- Start with [src/error.rs](https://github.com/leynos/wireframe/blob/issue-513-wireframeerror-e-should-derive-display-and-std-error-error/src/error.rs) to review the `std::error::Error` implementation and source chaining behaviour.
- Then review [src/testkit/result.rs](https://github.com/leynos/wireframe/blob/issue-513-wireframeerror-e-should-derive-display-and-std-error-error/src/testkit/result.rs) for the transparent `TestError::Wireframe` conversion.
- Finish with [tests/error_display.rs](https://github.com/leynos/wireframe/blob/issue-513-wireframeerror-e-should-derive-display-and-std-error-error/tests/error_display.rs) for the coverage around `WireframeError<()>` sources.

## Validation

- `make check-fmt`: passed
- `make lint`: passed
- `make test`: passed
- `coderabbit review --agent`: passed with zero findings after one rate-limit retry

## Notes

The originally suggested blanket `Display`/`Error` implementation plus a separate `WireframeError<()>` specialisation conflicts under Rust's coherence rules because upstream crates may add future trait implementations for `()`. This branch therefore uses a stable-compatible `Error` implementation for `WireframeError<E>` where `E: Debug + 'static`; `source()` exposes the nested `Io` and `Codec` errors that are always known to implement `Error`.

## Summary by Sourcery

Integrate WireframeError with std::error::Error while introducing a NoProtocolError default marker and tightening the public Result and response error surface.

New Features:
- Make WireframeError<E> implement std::error::Error when the protocol error type E is an Error, preserving protocol failures as sources.
- Introduce the NoProtocolError marker type and use WireframeError<NoProtocolError> as the crate-wide default error for Result<T> and response-related aliases.

Enhancements:
- Adjust Response and FrameStream aliases, examples, and fixtures to use the NoProtocolError-backed WireframeError default instead of unit.
- Allow TestError to use thiserror's transparent #[from] conversion for WireframeError while preserving its display prefix.
- Document the new error surface conventions, default protocol marker, and migration guidance in the library, user guide, and developer guide.

Build:
- Add the trybuild dev-dependency and wire up compile-time UI tests for the default WireframeError and Result<T> behavior.

Documentation:
- Update user, developer, migration, and crate root documentation to describe the NoProtocolError marker, the default WireframeError contract, and how protocol-specific error types participate in source chains.

Tests:
- Expand error_display tests to cover WireframeError source chaining, default/parameterized protocol types, async propagation, boxing as dyn Error, and TestError integration.
- Add trybuild-based UI tests to assert that the default Result<T> uses WireframeError<NoProtocolError> and does not accept unit protocol payloads by default.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai pause

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

~~~
Please address the comments from this code review:

## Overall Comments
- In `impl std::error::Error for WireframeError<E>`, the `E: Debug + 'static` bound looks stricter than needed since `source()` never exposes `E`; consider relaxing this back to `E: Debug` unless you specifically need `'static` for downstream usage like type-erased downcasting on the outer error.
- Switching `TestError::Wireframe` to `#[error(transparent)]` changes the user-visible error message format (dropping the `"wireframe error: "` prefix); if this change in Display output is not intentional, you may want to preserve the prefix via a custom `Display` or a separate variant.

## Individual Comments

### Comment 1
<location path="src/testkit/result.rs" line_range="24-25" />
<code_context>
     /// Root crate error surfaced while exercising a helper.
-    #[error("wireframe error: {0}")]
-    Wireframe(WireframeError),
+    #[error(transparent)]
+    Wireframe(#[from] WireframeError),
     /// Client-side error surfaced while exercising a helper.
     #[cfg(not(loom))]
</code_context>
<issue_to_address>
**issue (bug_risk):** Switching to `transparent` changes the public error message format and may affect callers relying on the previous prefix.

This variant used to emit messages like `"wireframe error: ..."`; with `#[error(transparent)]` the inner `WireframeError` message is now shown as-is. That can break consumers, logging, or error-matching that rely on the old prefix. If the prefix is important for log disambiguation or compatibility, consider keeping it in `WireframeError`’s `Display` (or not using `transparent` here) while still using `#[from]` for conversion.
</issue_to_address>

### Comment 2
<location path="tests/error_display.rs" line_range="53-51" />
<code_context>
+    );
+}
+
 #[test]
 fn wireframe_error_exposes_sources_for_io_and_codec() {
     let io = WireframeError::<ProtoErr>::from_io(io::Error::other("socket closed"));
</code_context>
<issue_to_address>
**suggestion (testing):** Consider also asserting `Display` output for the `WireframeError<()>` variants for consistency with other tests.

This test currently only validates `source()` behavior for the unit specialization, while the others in this file also assert on `Display`. To keep coverage consistent and catch formatting regressions, please add `to_string()` assertions for the `io`, `codec`, `protocol`, and `duplicate_route` `WireframeError<()>` variants in this test (or a nearby one).
</issue_to_address>

### Comment 3
<location path="src/error.rs" line_range="55" />
<code_context>
 impl<E> std::error::Error for WireframeError<E>
 where
-    E: std::fmt::Debug + std::error::Error + 'static,
+    E: std::fmt::Debug + 'static,
 {
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
</code_context>
<issue_to_address>
**issue (review_instructions):** Add tests that demonstrate and lock in the corrected WireframeError source behaviour and the relaxed error bound.

You changed the `Error` bound on `WireframeError<E>` and modified `source()` so that `Protocol` no longer returns an underlying error while `DuplicateRoute` and `Protocol` both return `None`. This is observable behaviour and appears to be a bug fix or at least a semantics change for error chaining. Add targeted tests (unit and behavioural) that fail with the old implementation and pass with this one, explicitly checking `source()` for each variant and the new bound behaviour.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*`

**Instructions:**
Bug fixes must be demonstrated by a test.

</details>
</issue_to_address>
~~~

File: tests/error_display.rs [hunk 1]
```diff
@@ -50,6 +50,40 @@ fn wireframe_error_messages() {
...
+        protocol.source().is_none(),
+        "unit protocol errors should not expose an error source"
+    );
+
+    let duplicate_route = WireframeError::<()>::DuplicateRoute(7);
+    assert!(
+        duplicate_route.source().is_none(),
+        "DuplicateRoute should not expose an error source"
+    );
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/error_display.rs:207]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/error_display.rs [hunk 2]
```diff
@@ -30,6 +33,35 @@ impl std::fmt::Display for ProtoErr {
...
+            let Some(codec_source) = source.downcast_ref::<CodecError>() else {
+                panic!("codec source should be wireframe::codec::CodecError");
+            };
+            assert!(
+                matches!(codec_source, CodecError::Framing(error) if error == &expected_error),
+                "codec source should preserve the original framing error"
+            );
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/error_display.rs]: @coderabbitai per CQRS, this should be separated into a fallible helper and a pure query that can be asserted on. Please provide an AI coding agent prompt for this fix.

[the remark this answers is not in the document]
Comment (leynos) [reply tests/error_display.rs]: @coderabbitai, no, like this:

~~~
I’d split it into:

1. **a fallible extractor** that turns the dynamic `Error` source into a small typed/normalised value;
2. **a pure predicate** that compares that value with `ExpectedSource`;
3. **the assertion wrapper**, which now just formats failure nicely.

Something like this:

```rust
#[derive(Debug, PartialEq, Eq)]
enum ActualSource<'a> {
    Io {
        message: String,
        kind: io::ErrorKind,
    },
    Codec {
        error: &'a FramingError,
    },
}

#[derive(Debug, PartialEq, Eq)]
enum SourceExtractionError {
    UnexpectedType,
    UnexpectedCodecVariant,
}

fn extract_source(source: &(dyn Error + 'static)) -> Result<ActualSource<'_>, SourceExtractionError> {
    if let Some(io_source) = source.downcast_ref::<io::Error>() {
        return Ok(ActualSource::Io {
            message: io_source.to_string(),
            kind: io_source.kind(),
        });
    }

    if let Some(codec_source) = source.downcast_ref::<CodecError>() {
        return match codec_source {
            CodecError::Framing(error) => Ok(ActualSource::Codec { error }),
            _ => Err(SourceExtractionError::UnexpectedCodecVariant),
        };
    }

    Err(SourceExtractionError::UnexpectedType)
}

fn source_matches(actual: &ActualSource<'_>, expected: &ExpectedSource) -> bool {
    match (actual, expected) {
        (
            ActualSource::Io {
                message: actual_message,
                kind: actual_kind,
            },
            ExpectedSource::Io {
                message: expected_message,
                kind: expected_kind,
            },
        ) => actual_kind == expected_kind && actual_message == expected_message,

        (
            ActualSource::Codec {
                error: actual_error,
            },
            ExpectedSource::Codec(expected_error),
        ) => *actual_error == expected_error,

        _ => false,
    }
}

fn assert_source_matches(source: &(dyn Error + 'static), expected: ExpectedSource) {
    let actual = extract_source(source)
        .expect("source should be a supported error source");

    assert!(
        source_matches(&actual, &expected),
        "source should match expected source\nactual: {actual:?}\nexpected: {expected:?}",
    );
}
```

That gives you a nice testing surface. You can now assert on the extractor independently:

```rust
let actual = extract_source(source).unwrap();

assert_eq!(
    actual,
    ActualSource::Io {
        message: "permission denied".to_owned(),
        kind: io::ErrorKind::PermissionDenied,
    }
);
```

And separately assert the pure comparison:

```rust
assert!(source_matches(&actual, &expected));
```

The one slight wrinkle is this bit:

```rust
ActualSource::Codec {
    error: &'a FramingError,
}
```

That avoids requiring the framing error to be `Clone`. If the framing error type is cheap and cloneable, you could instead make `ActualSource` fully owned:

```rust
Codec {
    error: FramingError,
}
```

and clone it during extraction. But I’d keep the borrowed version unless ownership buys you something. No need to make the test helper eat the sofa.
~~~

File: tests/error_display.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply tests/error_display.rs:207]: @coderabbitai feedback from the developer:

> Verified against current code. I did not make a commit.
> 
> The requested `WireframeError<()>: std::error::Error` compile-pass test is not implementable alongside the current blanket impl that preserves protocol sources:
> 
> ```rust
> impl<E> Error for WireframeError<E>where    E: Debug + Error + 'static
> ```
> 
> Adding `impl Error for WireframeError<()>` conflicts under Rust coherence rules, because upstream could add `impl Error for ()` in the future. I confirmed this by trying the minimal impl and running `cargo test compile_tests`; rustc rejected it with `E0119` for conflicting implementations.
> 
> `WireframeError<NoProtocolError>: Error` is already valid. `WireframeError<()>: Error` would require giving up the blanket protocol-source-preserving impl, or changing the public API shape more substantially.

File: src/error.rs [hunk 3]
```diff
@@ -52,14 +52,13 @@ impl<E: std::fmt::Debug> std::fmt::Display for WireframeError<E> {
...
-    E: std::fmt::Debug + std::error::Error + 'static,
+    E: std::fmt::Debug + 'static,
 {
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
         match self {
             Self::Io(error) => Some(error),
-            Self::Protocol(error) => Some(error),
             Self::Codec(error) => Some(error),
-            Self::DuplicateRoute(_) => None,
+            Self::DuplicateRoute(_) | Self::Protocol(_) => None,
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/error.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests omit coverage of a critical implementation detail: Protocol(NoProtocolError) must return None from source(). Line 116 constructs Protocol(()) but only asserts Display, not source(). | Add test that constructs Protocol(NoProtocolError) and asserts source() returns None, verifying the distinction between the blanket impl and specialization. |
| Module-Level Documentation | ❌ Error | src/testkit/result.rs and tests/stream\_end.rs have only single-line docstrings that lack substantive explanation of purpose, utility, function, and component relationships. | Expand both docstrings to explain purpose, utility/function, and relationships to testkit infrastructure and streaming response testing respectively. |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 76.19% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| User-Facing Documentation | ⚠️ Warning | PR introduces breaking API changes (default generic WireframeError<E = NoProtocolError>) but provides no user-facing documentation or migration guidance as flagged in review comments. | Add or update documentation (users-guide.md or new migration guide) explaining the default generic parameter change, its impact on Result, and required code updates for downstream consumers. |
| Developer Documentation | ⚠️ Warning | PR makes architectural API changes (NoProtocolError, std::error::Error integration) without updating developers-guide, roadmap, or migration documentation. | Update developers-guide with error implementation patterns; update v0-2-0-to-v0-3-0 migration guide to document NoProtocolError default and std::error::Error integration details. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Unit tests verify Display/source() but lack behavioural tests exercising the primary use-case: error wrapping in Box and async Result propagation. | Add integration tests demonstrating WireframeError wrapped as Box and propagated through async workflows to validate real-world error handling. |
| Architectural Complexity And Maintainability | ⚠️ Warning | Introduces `NoProtocolError` marker type instead of the suggested `E = ()` with two impls. Single-purpose, non-reusable abstraction adds API surface without reducing complexity. | Replace `E = NoProtocolError` default with `E = ()` and adopt the two-impl strategy from issue `#513`. Remove the `NoProtocolError` type and its re-exports. |

File: tests/error_display.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply tests/error_display.rs:207]: @coderabbitai the engineer notes:

> Verified against current code. No change was needed.
> 
> The requested compile-pass coverage already exists:
> 
> - `trybuild = "1"` is already in `Cargo.toml`.
> - `tests/compile_error.rs` has `compile_tests`.
> - `tests/ui/wireframe_result_default_no_protocol.rs` asserts `WireframeError<NoProtocolError>` can be passed to `&dyn std::error::Error`.
> - No `WireframeError<()>` pass test is present.
> 
> Validation passed:
> 
> - `cargo test compile_tests`
> 
> Working tree is clean; nothing to commit or push.

**Comment by leynos** [comment]
@sourcery-ai review

**Comment by leynos** [comment]
@coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

~~~
Please address the comments from this code review:

## Overall Comments
- The two `std::error::Error` impls for `WireframeError<E>` and `WireframeError<NoProtocolError>` now duplicate the `Io`/`Codec` handling; consider factoring out a small internal helper for source selection so the behavior doesn’t accidentally diverge over time.
- Now that `WireframeError` and `FrameStream` default to `NoProtocolError`, it may be clearer and more consistent to rely on the default type parameter (or explicitly use `NoProtocolError`) instead of `()` in places like `FrameStream<u8, ()>` and `ConnectionActor<_, ()>` unless you intentionally want the unit protocol type.

## Individual Comments

### Comment 1
<location path="src/error.rs" line_range="71" />
<code_context>
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
         match self {
             Self::Io(error) => Some(error),
-            Self::Protocol(error) => Some(error),
</code_context>
<issue_to_address>
**issue (review_instructions):** This new `match` has more than two branches, which should be extracted into a separate predicate/helper function per the guidelines.

The `source` implementation for `WireframeError<NoProtocolError>` introduces a `match` with three logical branches. To comply with the "conditionals with >2 branches" rule, consider extracting the branching logic into a dedicated helper/predicate function (e.g., `fn protocol_source(e: &WireframeError<NoProtocolError>) -> Option<&(dyn Error + 'static)>`) and call that from `source`. This keeps the function simpler and aligns with the stated complexity guidelines.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Move conditionals with >2 branches into a predicate function.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai please can you give me your thoughts on this suggestion?

Architectural Complexity And Maintainability ⚠️ Warning New NoProtocolError marker and cascading default changes across WireframeError, Response, FrameStream, ConnectionActor are disproportionate to the core problem solved. Implement Error directly for WireframeError<()>. Revert unnecessary default parameter changes. Solves `#513` without new abstractions.

**Comment by leynos** [comment]
@coderabbitai #532 has been closed as wontfix with rationale, and additional documentation has been added to this PR's changes explaining the motivations and intent.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review
